File: app/src/main/java/com/hiddenramblings/tagmo/nfctech/TagWriter.java
package com.hiddenramblings.tagmo.nfctech;

import android.content.Context;

import com.hiddenramblings.tagmo.R;
import com.hiddenramblings.tagmo.TagMo;
import com.hiddenramblings.tagmo.amiibo.KeyManager;
import com.hiddenramblings.tagmo.amiibo.PowerTagManager;
import com.hiddenramblings.tagmo.eightbit.io.Debug;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class TagWriter {

    public static void writeToTagRaw(
            NTAG215 mifare, byte[] tagData, boolean validateNtag
    ) throws Exception {
        final Context context = TagMo.getContext();
        TagUtils.validateNtag(mifare, tagData, validateNtag);
        TagReader.validateBlankTag(mifare);

        try {
            byte[][] pages = TagUtils.splitPages(tagData);
            writePages(mifare, 3, 129, pages);
            Debug.Info(TagWriter.class, R.string.data_write);
        } catch (Exception e) {
            throw new Exception(context.getString(R.string.error_data_write), e);
        }
        try {
            writePassword(mifare);
            Debug.Info(TagWriter.class, R.string.password_write);
        } catch (Exception e) {
            throw new Exception(context.getString(R.string.error_password_write), e);
        }
        try {
            writeLockInfo(mifare);
            Debug.Info(TagWriter.class, R.string.lock_write);
        } catch (Exception e) {
            throw new Exception(context.getString(R.string.error_lock_write), e);
        }
    }

    private static void writePages(
            NTAG215 tag, int pagestart, int pageend, byte[][] data
    ) throws IOException {
        for (int i = pagestart; i <= pageend; i++) {
            tag.writePage(i, data[i]);
            Debug.Info(TagWriter.class, R.string.write_page, String.valueOf(i));
        }
    }

    private static byte[] patchUid(byte[] uid, byte[] tagData) throws Exception {
        if (uid.length < 9) throw new IOException(TagMo.getContext()
                .getString(R.string.invalid_uid_length));

        byte[] patched = Arrays.copyOf(tagData, tagData.length);
        System.arraycopy(uid, 0, patched, 0x1d4, 8);
        patched[0] = uid[8];
        return patched;
    }

    private static void writePasswordLockInfo(NTAG215 mifare) throws Exception {
        try {
            writePassword(mifare);
            Debug.Info(TagWriter.class, R.string.password_write);
        } catch (Exception e) {
            throw new Exception(TagMo.getContext()
                    .getString(R.string.error_password_write), e);
        }
        try {
            writeLockInfo(mifare);
            Debug.Info(TagWriter.class, R.string.lock_write);
        } catch (Exception e) {
            throw new Exception(TagMo.getContext()
                    .getString(R.string.error_lock_write), e);
        }
    }

    public static void writeToTagAuto(
            NTAG215 mifare, byte[] tagData, KeyManager keyManager, boolean validateNtag
    ) throws Exception {
        byte[] idPages = mifare.readPages(0);
        if (null == idPages  || idPages.length != NfcByte.PAGE_SIZE * 4)
            throw new IOException(TagMo.getContext()
                    .getString(R.string.fail_read_size));

        boolean isPowerTag = TagUtils.isPowerTag(mifare);
        Debug.Info(TagWriter.class, R.string.power_tag_verify, String.valueOf(isPowerTag));

        try {
            tagData = keyManager.decrypt(tagData);
        } catch (Exception e) {
            throw new Exception(e);
        }

        if (isPowerTag) {
            // use a pre-determined static id for Power Tag
            tagData = patchUid(NfcByte.POWERTAG_IDPAGES, tagData);
        } else {
            tagData = patchUid(idPages, tagData);
        }
        tagData = keyManager.encrypt(tagData);

        Debug.Info(TagWriter.class, TagUtils.bytesToHex(tagData));

        if (!isPowerTag) {
            TagUtils.validateNtag(mifare, tagData, validateNtag);
            try {
                TagReader.validateBlankTag(mifare);
            } catch (IOException e) {
                throw new IOException(e);
            }
        }

        if (isPowerTag) {
            byte[] oldid = mifare.getTag().getId();
            if (null == oldid  || oldid.length != 7)
                throw new Exception(TagMo.getContext()
                        .getString(R.string.fail_read_uid));

            Debug.Info(TagWriter.class, R.string.old_uid, TagUtils.bytesToHex(oldid));

            byte[] page10 = mifare.readPages(0x10);
            Debug.Info(TagWriter.class, R.string.page_ten, TagUtils.bytesToHex(page10));

            String page10bytes = TagUtils.bytesToHex(new byte[]{page10[0], page10[3]});

            byte[] ptagKeySuffix = PowerTagManager.getPowerTagKey(oldid, page10bytes);
            byte[] ptagKey = TagUtils.hexToByteArray(NfcByte.POWERTAG_KEY);
            System.arraycopy(ptagKeySuffix, 0, ptagKey, 8, 8);

            Debug.Info(TagWriter.class, R.string.ptag_key, TagUtils.bytesToHex(ptagKey));

            mifare.transceive(NfcByte.POWERTAG_WRITE);
            mifare.transceive(ptagKey);

            if (!(idPages[0] == (byte) 0xFF && idPages[1] == (byte) 0xFF))
                doAuth(mifare);
        }

        byte[][] pages = TagUtils.splitPages(tagData);
        if (isPowerTag) {
            byte[] zeropage = TagUtils.hexToByteArray("00000000");
            mifare.writePage(0x86, zeropage); //PACK
            writePages(mifare, 0x01, 0x84, pages);
            mifare.writePage(0x85, zeropage); //PWD
            mifare.writePage(0x00, pages[0]); //UID
            mifare.writePage(0x00, pages[0]); //UID
        } else {
            try {
                writePages(mifare, 3, 129, pages);
                Debug.Info(TagWriter.class, R.string.data_write);
            } catch (Exception e) {
                throw new Exception(TagMo.getContext().getString(R.string.error_data_write), e);
            }
            writePasswordLockInfo(mifare);
        }
    }

    public static void writeEliteAuto(
            NTAG215 mifare, byte[] tagData, KeyManager keyManager, int active_bank
    ) throws Exception {
        if (doEliteAuth(mifare, mifare.fastRead(0, 0))) {
            tagData = keyManager.decrypt(tagData);
            // tagData = patchUid(mifare.readPages(0), tagData);
            tagData = keyManager.encrypt(tagData);
            boolean write = mifare.amiiboFastWrite(0, active_bank, tagData);
            if (!write) write = mifare.amiiboWrite(0, active_bank, tagData);
            if (!write) throw new IOException(TagMo.getContext()
                    .getString(R.string.error_elite_write));
        } else {
            throw new Exception(TagMo.getContext()
                    .getString(R.string.error_elite_auth));
        }
    }

    public static void restoreTag(
            NTAG215 mifare, byte[] tagData, boolean ignoreUid,
            KeyManager keyManager, boolean validateNtag
    ) throws Exception {
        if (!ignoreUid)
            TagUtils.validateNtag(mifare, tagData, validateNtag);
        else {
            byte[] liveData = TagReader.readFromTag(mifare);
            if (!TagUtils.compareRange(liveData, tagData, 9)) {
                // restoring to different tag: transplant mii and appdata to livedata and re-encrypt
                liveData = keyManager.decrypt(liveData);
                tagData = keyManager.decrypt(tagData);
                System.arraycopy(tagData, 0x08, liveData, 0x08, 0x1B4 - 0x08);
                /* TODO: Verify that 0x1B4 should not be 0x1D4 */
                tagData = keyManager.encrypt(liveData);
            }
        }

        doAuth(mifare);
        byte[][] pages = TagUtils.splitPages(tagData);
        writePages(mifare, 4, 12, pages);
        writePages(mifare, 32, 129, pages);
    }

    /**
     * Remove the checksum bytes from the first two pages to get the actual uid
     */
    private static byte[] uidFromPages(byte[] pages0_1) {
        if (pages0_1.length < 8) return null;

        byte[] key = new byte[7];
        key[0] = pages0_1[0];
        key[1] = pages0_1[1];
        key[2] = pages0_1[2];
        key[3] = pages0_1[4];
        key[4] = pages0_1[5];
        key[5] = pages0_1[6];
        key[6] = pages0_1[7];
        return key;
    }

    private static byte[] keygen(byte[] uuid) {
        // from AmiiManage (GPL)
        byte[] key = new byte[4];
        int[] uuid_to_ints = new int[uuid.length];

        for (int i = 0; i < uuid.length; i++)
            uuid_to_ints[i] = (0xFF & uuid[i]);

        if (uuid.length == 7) {
            key[0] = ((byte) (0xFF & (0xAA ^ (uuid_to_ints[1] ^ uuid_to_ints[3]))));
            key[1] = ((byte) (0xFF & (0x55 ^ (uuid_to_ints[2] ^ uuid_to_ints[4]))));
            key[2] = ((byte) (0xFF & (0xAA ^ (uuid_to_ints[3] ^ uuid_to_ints[5]))));
            key[3] = ((byte) (0xFF & (0x55 ^ (uuid_to_ints[4] ^ uuid_to_ints[6]))));
            return key;
        }

        return null;
    }

    @SuppressWarnings("ConstantConditions")
    private static void doAuth(NTAG215 tag) throws Exception {
        byte[] pages0_1 = tag.readPages(0);

        if (null == pages0_1  || pages0_1.length != NfcByte.PAGE_SIZE * 4)
            throw new IOException(TagMo.getContext().getString(R.string.fail_read));

        byte[] uid = uidFromPages(pages0_1);
        byte[] password = keygen(uid);

        Debug.Info(TagWriter.class, R.string.password, TagUtils.bytesToHex(password));

        byte[] auth = new byte[]{
                (byte) 0x1B,
                password[0],
                password[1],
                password[2],
                password[3]
        };
        byte[] response = tag.transceive(auth);
        if (null == response )
            throw new Exception(TagMo.getContext().getString(R.string.error_auth_null));
        String respStr = TagUtils.bytesToHex(response);
        Debug.Info(TagWriter.class, R.string.auth_response, respStr);
        if (!"8080".equals(respStr)) {
            throw new Exception(TagMo.getContext().getString(R.string.fail_auth));
        }
    }

    private static boolean doEliteAuth(NTAG215 tag, byte[] password) {
        if (null == password  || password.length != 4) {
            return false;
        }
        byte[] req = new byte[5];
        req[0] = NfcByte.CMD_PWD_AUTH;
        try {
            System.arraycopy(password, 0, req, 1, 4);
            password = tag.transceive(req);
        } catch (Exception e) {
            return false;
        }
        if (null == password  || password.length != 2) {
            return false;
        }
        return password[0] == Byte.MIN_VALUE && password[1] == Byte.MIN_VALUE;
    }

    @SuppressWarnings("ConstantConditions")
    private static void writePassword(NTAG215 tag) throws IOException {
        byte[] pages0_1 = tag.readPages(0);

        if (null == pages0_1  || pages0_1.length != NfcByte.PAGE_SIZE * 4)
            throw new IOException(TagMo.getContext().getString(R.string.fail_read));

        byte[] uid = uidFromPages(pages0_1);
        byte[] password = keygen(uid);

        Debug.Info(TagWriter.class, R.string.password, TagUtils.bytesToHex(password));

        Debug.Info(TagWriter.class, R.string.write_pack);
        tag.writePage(0x86, new byte[]{(byte) 0x80, (byte) 0x80, (byte) 0, (byte) 0});

        Debug.Info(TagWriter.class, R.string.write_pwd);
        tag.writePage(0x85, password);
    }

    private static void writeLockInfo(NTAG215 tag) throws IOException {
        byte[] pages = tag.readPages(0);

        if (null == pages  || pages.length != NfcByte.PAGE_SIZE * 4)
            throw new IOException(TagMo.getContext().getString(R.string.fail_read));

        tag.writePage(2, new byte[]{pages[2 * NfcByte.PAGE_SIZE],
                pages[(2 * NfcByte.PAGE_SIZE) + 1], (byte) 0x0F, (byte) 0xE0}); // lock bits
        tag.writePage(130, new byte[]{(byte) 0x01, (byte) 0x00, (byte) 0x0F, (byte) 0x00});
        // dynamic lock bits. should the last bit be 0xBD according to the nfc docs though:
        // Remark: Set all bits marked with RFUI to 0, when writing to the dynamic lock bytes.
        tag.writePage(131, new byte[]{(byte) 0x00, (byte) 0x00, (byte) 0x00, (byte) 0x04}); // config
        tag.writePage(132, new byte[]{(byte) 0x5F, (byte) 0x00, (byte) 0x00, (byte) 0x00}); // config
    }

    public static void wipeBankData(NTAG215 mifare, int active_bank)  throws Exception {
        if (doEliteAuth(mifare, mifare.fastRead(0, 0))) {
            byte[] tagData = TagUtils.hexToByteArray(new String(
                    new char[1080]).replace("\u0000", "F"));
            boolean write = mifare.amiiboFastWrite(0, active_bank, tagData);
            if (!write) write = mifare.amiiboWrite(0, active_bank, tagData);
            if (write) {
                byte[] result = new byte[8];
                System.arraycopy(tagData, 84, result, 0, result.length);
                Debug.Info(TagWriter.class, TagUtils.bytesToHex(result));
            } else {
                throw new Exception(TagMo.getContext()
                        .getString(R.string.error_elite_write));
            }
        } else {
            throw new Exception(TagMo.getContext()
                    .getString(R.string.error_elite_write));
        }
    }

    public static boolean updateFirmware(NTAG215 tag) throws Exception {
        final Context context = TagMo.getContext();
        byte[] response = new byte[1];
        response[0] = (byte) 0xFFFF;
        tag.initFirmware();
        tag.getVersion(true);
        try {
            BufferedReader br = new BufferedReader(new InputStreamReader(
                    context.getResources().openRawResource(R.raw.firmware)));
            while (true) {
                String strLine = br.readLine();
                if (null == strLine) {
                    break;
                }
                String[] parts = strLine.replaceAll("\\s+", " ").split(" ");
                int i;
                if (parts.length < 1) {
                    break;
                } else if (parts[0].equals("C-APDU")) {
                    byte[] apdu_buf = new byte[(parts.length - 1)];
                    for (i = 1; i < parts.length; i++) {
                        apdu_buf[i - 1] = TagUtils.hexToByte(parts[i]);
                    }
                    int sz = apdu_buf[4] & 0xFF;
                    byte[] iso_cmd = new byte[sz];
                    if (apdu_buf[4] + 5 <= apdu_buf.length && apdu_buf[4] <= iso_cmd.length) {
                        for (i = 0; i < sz; i++) {
                            iso_cmd[i] = apdu_buf[i + 5];
                        }
                        boolean done = false;
                        for (i = 0; i < 10; i++) {
                            response = tag.transceive(iso_cmd);
                            if (null != response) {
                                done = true;
                                break;
                            }
                        }
                        if (!done) {
                            throw new Exception(context.getString(R.string.firmware_failed, 1));
                        }
                    }
                    return false;
                } else if (parts[0].equals("C-RPDU")) {
                    byte[] rpdu_buf = new byte[(parts.length - 1)];
                    if (response.length != parts.length - 3) {
                        throw new Exception(context.getString(R.string.firmware_failed, 2));
                    }
                    for (i = 1; i < parts.length; i++) {
                        rpdu_buf[i - 1] = TagUtils.hexToByte(parts[i]);
                    }
                    for (i = 0; i < rpdu_buf.length - 2; i++) {
                        if (rpdu_buf[i] != response[i]) {
                            throw new Exception(context.getString(R.string.firmware_failed, 3));
                        }
                    }
                } /* else if (!parts[0].equals("RESET") && parts[0].equals("LOGIN")) { } */
            }
            br.close();
            return true;
        } catch (IOException e) {
            throw new Exception(context.getString(R.string.firmware_failed, 4));
        }
    }
}

File: app/src/main/java/com/hiddenramblings/tagmo/nfctech/TagWriter.kt
package com.hiddenramblings.tagmo.nfctech

import com.hiddenramblings.tagmo.R
import com.hiddenramblings.tagmo.TagMo
import com.hiddenramblings.tagmo.amiibo.KeyManager
import com.hiddenramblings.tagmo.amiibo.PowerTagManager
import com.hiddenramblings.tagmo.eightbit.io.Debug
import com.hiddenramblings.tagmo.nfctech.TagArray.isPowerTag
import com.hiddenramblings.tagmo.nfctech.TagArray.toDecryptedTag
import com.hiddenramblings.tagmo.nfctech.TagArray.toHex
import com.hiddenramblings.tagmo.nfctech.TagArray.toHexByteArray
import com.hiddenramblings.tagmo.nfctech.TagArray.toTagArray
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.util.*

object TagWriter {
    @Throws(Exception::class)
    private fun splitPages(data: ByteArray): Array<ByteArray?> {
        if (data.size < NfcByte.TAG_DATA_SIZE)
            throw IOException(TagMo.appContext.getString(
                R.string.invalid_data_size, data.size, NfcByte.TAG_DATA_SIZE
            ))
        val pages = arrayOfNulls<ByteArray>(data.size / NfcByte.PAGE_SIZE)
        var i = 0
        var j = 0
        while (i < data.size) {
            pages[j] = data.copyOfRange(i, i + NfcByte.PAGE_SIZE)
            i += NfcByte.PAGE_SIZE
            j++
        }
        return pages
    }

    @Throws(Exception::class)
    fun writeToTagRaw(
        mifare: NTAG215, tagData: ByteArray, validateNtag: Boolean
    ) {
        val context = TagMo.appContext
        TagArray.validateNtag(mifare, tagData, validateNtag)
        TagReader.validateBlankTag(mifare)
        try {
            val pages = splitPages(tagData)
            writePages(mifare, 3, 129, pages)
            Debug.verbose(TagWriter::class.java, R.string.data_write)
        } catch (e: Exception) {
            throw Exception(context.getString(R.string.error_data_write), e)
        }
        try {
            writePassword(mifare)
            Debug.verbose(TagWriter::class.java, R.string.password_write)
        } catch (e: Exception) {
            throw Exception(context.getString(R.string.error_password_write), e)
        }
        try {
            writeLockInfo(mifare)
            Debug.verbose(TagWriter::class.java, R.string.lock_write)
        } catch (e: Exception) {
            throw Exception(context.getString(R.string.error_lock_write), e)
        }
    }

    @Throws(IOException::class)
    private fun writePages(
        tag: NTAG215, pagestart: Int, pageend: Int, data: Array<ByteArray?>
    ) {
        for (i in pagestart..pageend) {
            tag.writePage(i, data[i]!!)
            Debug.verbose(TagWriter::class.java, R.string.write_page, i.toString())
        }
    }

    @Throws(Exception::class)
    private fun patchUid(uid: ByteArray, tagData: ByteArray): ByteArray {
        if (uid.size < 9) throw IOException(TagMo.appContext.getString(R.string.invalid_uid_length))
        val patched = tagData.copyOf(tagData.size)
        System.arraycopy(uid, 0, patched, 0x1d4, 8)
        patched[0] = uid[8]
        return patched
    }

    @Throws(Exception::class)
    private fun writePasswordLockInfo(mifare: NTAG215) {
        try {
            writePassword(mifare)
            Debug.verbose(TagWriter::class.java, R.string.password_write)
        } catch (e: Exception) {
            throw Exception(TagMo.appContext.getString(R.string.error_password_write), e)
        }
        try {
            writeLockInfo(mifare)
            Debug.verbose(TagWriter::class.java, R.string.lock_write)
        } catch (e: Exception) {
            throw Exception(TagMo.appContext.getString(R.string.error_lock_write), e)
        }
    }

    @Throws(Exception::class)
    fun writeToTagAuto(
        mifare: NTAG215, tagData: ByteArray, keyManager: KeyManager, validateNtag: Boolean
    ) {
        val idPages = mifare.readPages(0)
        if (null == idPages || idPages.size != NfcByte.PAGE_SIZE * 4)
            throw IOException(TagMo.appContext.getString(R.string.fail_read_size))
        val isPowerTag = mifare.isPowerTag
        Debug.verbose(TagWriter::class.java, R.string.power_tag_verify, isPowerTag.toString())
        var writeData = tagData.toTagArray().toDecryptedTag(keyManager)
        writeData = if (isPowerTag) {
            // use a pre-determined static id for Power Tag
            patchUid(NfcByte.POWERTAG_IDPAGES, writeData)
        } else {
            patchUid(idPages, writeData)
        }
        writeData = keyManager.encrypt(writeData)
        Debug.verbose(TagWriter::class.java, writeData.toHex())
        if (!isPowerTag) {
            TagArray.validateNtag(mifare, writeData, validateNtag)
            try {
                TagReader.validateBlankTag(mifare)
            } catch (e: IOException) {
                throw IOException(e)
            }
        }
        if (isPowerTag) {
            val oldid = mifare.tag!!.id
            if (null == oldid || oldid.size != 7)
                throw Exception(TagMo.appContext.getString(R.string.fail_read_uid))
            Debug.verbose(TagWriter::class.java, R.string.old_uid, oldid.toHex())
            val page10 = mifare.readPages(0x10)
            page10?.let { Debug.verbose(TagWriter::class.java, R.string.page_ten, it.toHex()) }
            val page10bytes = byteArrayOf(page10?.get(0) ?: 0, page10?.get(3) ?: 0).toHex()
            val ptagKeySuffix = PowerTagManager.getPowerTagKey(oldid, page10bytes)
            val ptagKey = NfcByte.POWERTAG_KEY.toHexByteArray()
            System.arraycopy(ptagKeySuffix, 0, ptagKey, 8, 8)
            Debug.verbose(TagWriter::class.java, R.string.ptag_key, ptagKey.toHex())
            mifare.transceive(NfcByte.POWERTAG_WRITE)
            mifare.transceive(ptagKey)
            if (!(idPages[0] == 0xFF.toByte() && idPages[1] == 0xFF.toByte())) doAuth(mifare)
        }
        val pages = splitPages(writeData)
        if (isPowerTag) {
            val zeropage = TagArray.hexToByteArray("00000000")
            mifare.writePage(0x86, zeropage) //PACK
            writePages(mifare, 0x01, 0x84, pages)
            mifare.writePage(0x85, zeropage) //PWD
            mifare.writePage(0x00, pages[0]!!) //UID
            mifare.writePage(0x00, pages[0]!!) //UID
        } else {
            try {
                writePages(mifare, 3, 129, pages)
                Debug.verbose(TagWriter::class.java, R.string.data_write)
            } catch (e: Exception) {
                throw Exception(TagMo.appContext.getString(R.string.error_data_write), e)
            }
            writePasswordLockInfo(mifare)
        }
    }

    @Throws(Exception::class)
    fun writeEliteAuto(
        mifare: NTAG215, tagData: ByteArray?, bankNumber: Int
    ) {
        if (doEliteAuth(mifare, mifare.fastRead(0, 0))) {
            var write = mifare.amiiboFastWrite(0, bankNumber, tagData)
            if (!write) write = mifare.amiiboWrite(0, bankNumber, tagData)
            if (!write) throw IOException(TagMo.appContext.getString(R.string.error_elite_write))
        } else {
            throw Exception(TagMo.appContext.getString(R.string.error_elite_auth))
        }
    }

    @Throws(Exception::class)
    fun restoreTag(
        mifare: NTAG215, tagData: ByteArray, ignoreUid: Boolean,
        keyManager: KeyManager, validateNtag: Boolean
    ) {
        var restoreData = tagData.toTagArray()
        if (!ignoreUid) TagArray.validateNtag(mifare, restoreData, validateNtag) else {
            var liveData = TagReader.readFromTag(mifare)
            if (!TagArray.compareRange(liveData, restoreData, 9)) {
                // restoring to different tag: transplant mii and appdata to livedata and re-encrypt
                liveData = keyManager.decrypt(liveData)
                restoreData = keyManager.decrypt(restoreData)
                System.arraycopy(restoreData, 0x08, liveData, 0x08, 0x1B4 - 0x08)
                /* TODO: Verify that 0x1B4 should not be 0x1D4 */
                restoreData = keyManager.encrypt(liveData)
            }
        }
        doAuth(mifare)
        val pages = splitPages(restoreData)
        writePages(mifare, 4, 12, pages)
        writePages(mifare, 32, 129, pages)
    }

    /**
     * Remove the checksum bytes from the first two pages to get the actual uid
     */
    private fun uidFromPages(pages0_1: ByteArray): ByteArray? {
        if (pages0_1.size < 8) return null
        return byteArrayOf(
            pages0_1[0],
            pages0_1[1],
            pages0_1[2],
            pages0_1[4],
            pages0_1[5],
            pages0_1[6],
            pages0_1[7]
        )
    }

    private fun keygen(uuid: ByteArray?): ByteArray? {
        // from AmiiManage (GPL)
        if (null == uuid) return null
        val uuidIntArray = IntArray(uuid.size)
        uuid.forEachIndexed { i, byte ->
            uuidIntArray[i] = 0xFF and byte.toInt()
        }
        if (uuid.size == 7) {
            return byteArrayOf(
                (0xFF and (0xAA xor (uuidIntArray[1] xor uuidIntArray[3]))).toByte(),
                (0xFF and (0x55 xor (uuidIntArray[2] xor uuidIntArray[4]))).toByte(),
                (0xFF and (0xAA xor (uuidIntArray[3] xor uuidIntArray[5]))).toByte(),
                (0xFF and (0x55 xor (uuidIntArray[4] xor uuidIntArray[6]))).toByte()
            )
        }
        return null
    }

    @Throws(Exception::class)
    private fun doAuth(tag: NTAG215) {
        val pages01 = tag.readPages(0)
        if (null == pages01 || pages01.size != NfcByte.PAGE_SIZE * 4)
            throw IOException(TagMo.appContext.getString(R.string.fail_read))
        val uid = uidFromPages(pages01)
        val password = keygen(uid)
        Debug.verbose(TagWriter::class.java, R.string.password, password!!.toHex())
        val auth = byteArrayOf(
            0x1B.toByte(),
            password[0],
            password[1],
            password[2],
            password[3]
        )
        val response = tag.transceive(auth)
            ?: throw Exception(TagMo.appContext.getString(R.string.error_auth_null))
        val respStr = response.toHex()
        Debug.verbose(TagWriter::class.java, R.string.auth_response, respStr)
        if ("8080" != respStr) {
            throw Exception(TagMo.appContext.getString(R.string.fail_auth))
        }
    }

    private fun doEliteAuth(tag: NTAG215, password: ByteArray?): Boolean {
        var passBytes: ByteArray? = password
        if (null == passBytes || passBytes.size != 4) return false
        val req = ByteArray(5)
        req[0] = NfcByte.CMD_PWD_AUTH.toByte()
        passBytes = try {
            System.arraycopy(passBytes, 0, req, 1, 4)
            tag.transceive(req)
        } catch (e: Exception) {
            return false
        }
        return if (null == passBytes || passBytes.size != 2) {
            false
        } else {
            passBytes[0] == Byte.MIN_VALUE && passBytes[1] == Byte.MIN_VALUE
        }
    }

    @Throws(IOException::class)
    private fun writePassword(tag: NTAG215) {
        val pages01 = tag.readPages(0)
        if (null == pages01 || pages01.size != NfcByte.PAGE_SIZE * 4)
            throw IOException(TagMo.appContext.getString(R.string.fail_read))
        val uid = uidFromPages(pages01)
        val password = keygen(uid)
        Debug.verbose(TagWriter::class.java, R.string.password, password!!.toHex())
        Debug.verbose(TagWriter::class.java, R.string.write_pack)
        tag.writePage(0x86, byteArrayOf(0x80.toByte(), 0x80.toByte(), 0.toByte(), 0.toByte()))
        Debug.verbose(TagWriter::class.java, R.string.write_pwd)
        tag.writePage(0x85, password)
    }

    @Throws(IOException::class)
    private fun writeLockInfo(tag: NTAG215) {
        val pages = tag.readPages(0)
        if (null == pages || pages.size != NfcByte.PAGE_SIZE * 4)
            throw IOException(TagMo.appContext.getString(R.string.fail_read))
        tag.writePage(
            2, byteArrayOf(
                pages[2 * NfcByte.PAGE_SIZE],
                pages[2 * NfcByte.PAGE_SIZE + 1], 0x0F.toByte(), 0xE0.toByte()
            )
        ) // lock bits
        tag.writePage(130, byteArrayOf(0x01.toByte(), 0x00.toByte(), 0x0F.toByte(), 0x00.toByte()))
        // dynamic lock bits. should the last bit be 0xBD according to the nfc docs though:
        // Remark: Set all bits marked with RFUI to 0, when writing to the dynamic lock bytes.
        tag.writePage(
            131, byteArrayOf(0x00.toByte(), 0x00.toByte(), 0x00.toByte(), 0x04.toByte())
        ) // config
        tag.writePage(
            132, byteArrayOf(0x5F.toByte(), 0x00.toByte(), 0x00.toByte(), 0x00.toByte())
        ) // config
    }

    @Throws(Exception::class)
    fun wipeBankData(mifare: NTAG215, bankNumber: Int) {
        if (doEliteAuth(mifare, mifare.fastRead(0, 0))) {
            val tagData = String(CharArray(1080))
                    .replace("\u0000", "F").toHexByteArray()
            var write = mifare.amiiboFastWrite(0, bankNumber, tagData)
            if (!write) write = mifare.amiiboWrite(0, bankNumber, tagData)
            if (write) {
                val result = ByteArray(8)
                System.arraycopy(tagData, 84, result, 0, result.size)
                Debug.verbose(TagWriter::class.java, result.toHex())
            } else {
                throw Exception(TagMo.appContext.getString(R.string.error_elite_write))
            }
        } else {
            throw Exception(TagMo.appContext.getString(R.string.error_elite_write))
        }
    }

    private fun hexToByte(hex: String): Byte {
        var ret = 0.toByte()
        val hi = hex[0].code.toByte()
        val lo = hex[1].code.toByte()
        when {
            hi >= NfcByte.CMD_READ && hi <= NfcByte.CMD_READ_CNT ->
                ret = (hi - 0x30 shl 4).toByte()
            hi >= 0x41.toByte() && hi <= NfcByte.N2_LOCK ->
                ret = (hi - 0x41 + 0x0A shl 4).toByte()
            hi >= 0x61.toByte() && hi <= 0x66.toByte() ->
                ret = (hi - 0x61 + 0x0A shl 4).toByte()
        }
        return when {
            lo >= NfcByte.CMD_READ && lo <= NfcByte.CMD_READ_CNT ->
                (lo - 0x30 or ret.toInt()).toByte()
            lo >= 0x41.toByte() && lo <= NfcByte.N2_LOCK ->
                (lo - 0x41 + 0x0A or ret.toInt()).toByte()
            lo < 0x61.toByte() || lo > 0x66.toByte() -> ret
            else -> (lo - 0x61 + 0x0A or ret.toInt()).toByte()
        }
    }

    @Throws(Exception::class)
    fun updateFirmware(tag: NTAG215?): Boolean {
        if (null == tag) return false
        val context = TagMo.appContext
        var response: ByteArray? = ByteArray(1)
        response?.set(0, 0xFFFF.toByte())
        tag.initFirmware()
        tag.getVersion(true)
        return try {
            BufferedReader(InputStreamReader(
                context.resources.openRawResource(R.raw.firmware)
            )).use { br ->
                while (true) {
                    val strLine = br.readLine() ?: break
                    val parts = strLine.replace(
                        "\\s+".toRegex(), " "
                    ).split(" ".toRegex()).toTypedArray()
                    var i: Int
                    if (parts.isEmpty()) {
                        break
                    } else if (parts[0] == "C-APDU") {
                        val apduBuf = ByteArray(parts.size - 1)
                        i = 1
                        while (i < parts.size) {
                            apduBuf[i - 1] = hexToByte(parts[i])
                            i++
                        }
                        val sz: Int = apduBuf[4].toInt() and 0xFF
                        val isoCmd = ByteArray(sz)
                        if (apduBuf[4] + 5 <= apduBuf.size && apduBuf[4] <= isoCmd.size) {
                            i = 0
                            while (i < sz) {
                                isoCmd[i] = apduBuf[i + 5]
                                i++
                            }
                            var done = false
                            i = 0
                            while (i < 10) {
                                response = tag.transceive(isoCmd)
                                if (null != response) {
                                    done = true
                                    break
                                }
                                i++
                            }
                            if (!done)
                                throw Exception(context.getString(R.string.firmware_failed, 1))
                        }
                        return false
                    } else if (parts[0] == "C-RPDU") {
                        val rpduBuf = ByteArray(parts.size - 1)
                        if (response?.size != parts.size - 3)
                            throw Exception(context.getString(R.string.firmware_failed, 2))
                        i = 1
                        while (i < parts.size) {
                            rpduBuf[i - 1] = hexToByte(parts[i])
                            i++
                        }
                        i = 0
                        while (i < rpduBuf.size - 2) {
                            if (response?.get(i) != rpduBuf[i])
                                throw Exception(context.getString(R.string.firmware_failed, 3))
                            i++
                        }
                    } /* else if (!parts[0].equals("RESET") && parts[0].equals("LOGIN")) { } */
                }
                true
            }
        } catch (e: IOException) {
            throw Exception(context.getString(R.string.firmware_failed, 4))
        }
    }
}